Fix CMake syntax to enable memory analyzer tests

tautschnig opened the pull request:
"option" requires the help text as second, not as third argument. This
resulted in memory analyzer tests being disabled no matter what the OS
being built on.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Fix CMake syntax to enable memory analyzer tests

"option" requires the help text as second, not as third argument. This
resulted in memory analyzer tests being disabled no matter what the OS
being built on.

tautschnig committed:
Enable extract_type_header regression tests

This test isn't strictly dependent on memory-analyzer being available,
but is only of interest if that is the case.

TGWDB commented on unit/memory-analyzer/gdb_api.cpp:
```diff
@@ -18,13 +18,61 @@ Author: Malte Mues <mail.mues@gmail.com>
+          "\n"
+          "char *s = \"abc\";\n"
+          "int *p;\n"
+          "void *vp;\n"
+          "int *np = 0;\n"
+          "void *vp_string;\n"
+          "\n"
+          "void checkpoint()\n"
+          "{\n"
+          "}\n"
+          "void checkpoint2()\n"
+          "{\n"
+          "}\n"
+          "\n"
+          "void func()\n"
+          "{\n"
+          "  checkpoint2();\n"
+          "}\n"
+          "\n"
+          "int main()\n"
+          "{\n"
+          "  x = 8;\n"
+          "  y = 2.5;\n"
+          "  z = 'c';\n"
+          "\n"
+          "  p = &x;\n"
+          "  vp = (void *)&x;\n"
+          "  vp_string = s;\n"
+          "\n"
+          "  checkpoint();\n"
+          "\n"
+          "  return 0;\n"
+          "}\n";
+    of.close();
+
+    REQUIRE(run("gcc", {"gcc", "-g", "-o", compiled(), tmp()}) == 0);
+  }
+
+  temporary_filet compiled;
+};
```
This seems unnecessarily ugly, is there a reason to do it this way?

tautschnig replied:
If, as was the case before, we keep the contents in a separate file, then this file must either be included in the compiled unit test binary (e.g., via `#include`), or else the compiled unit test binary necessarily relies on the source tree being present as well. It seems reasonable to assume that the source tree will be present, but then also the exact location of the file must be known.

I made these changes as I found the above conditions not to be satisfied in my ad-hoc testing for I wasn't aware where the unit test had to be run from to make things work. That being said, we have the same problem in other unit tests. So maybe it's not all that important to fix if people are concerned about this approach.

TGWDB replied:
I understand the motivation, I was more considering that there might be a prettier way to have the text inside the file.

TGWDB commented on unit/memory-analyzer/gdb_api.cpp:
```diff
@@ -67,9 +115,34 @@ void gdb_api_internals_test()
   SECTION("read a line from an input stream")
   {
+    temporary_filet tmp("input", ".txt");
+    std::ofstream of(tmp().c_str());
+    REQUIRE(of.is_open());
+
+    of << "abc\n"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\n"
+          "xyz";
+    of.close();
```
Again, is this the best way to do this?

tautschnig replied:
I've modified this to use `ansi-c/file_converter` and `#include` them instead.

tautschnig committed:
memory analyzer: fail when gdb cannot be started

When gdb wasn't available, the forked child would not terminate and the
parent would wait forever.

tautschnig committed:
Memory analyzer unit tests: embed source files

Make the unit tests less fragile by not relying on a particular
directory that they are being run from. Test files had to be loaded from
a specific location; instead, place them in the source code and generate
temporary files at runtime.
